File: gateway/ts/src/core/protocol.ts
export const CONTROL_VERSION = 1;
export const CONTRACT_VERSION = 1;

export type ErrorCode =
  | "CANCELLED"
  | "RESOURCE_EXHAUSTED"
  | "UNKNOWN_ROUTE"
  | "OWNER_UNAVAILABLE"
  | "INVALID_ARGUMENT"
  | "CONCURRENCY_FULL"
  | "TIMEOUT"
  | "UNAUTHORIZED"
  | "HANDLER_ERROR"
  | "CONFLICT"
  | "INCOMPATIBLE"
  | "WRONG_METHOD_KIND";
export interface GatewayError {
  code: ErrorCode;
  message: string;
}
export type Result<T> = { ok: true; value: T } | { ok: false; error: GatewayError };
export const failure = (code: ErrorCode, message: string): Result<never> => ({ ok: false, error: { code, message } });
export const success = <T>(value: T): Result<T> => ({ ok: true, value });

// Exceptions are local convenience only; transport always sends the plain Result union.
export class GatewayFailure extends Error {
  constructor(readonly detail: GatewayError) {
    super(detail.message);
  }
}
export function unwrap<T>(result: Result<T>): T {
  if (!result.ok) throw new GatewayFailure(result.error);
  return result.value;
}
export function invalid(message: string): never {
  throw new GatewayFailure({ code: "INVALID_ARGUMENT", message });
}
export function validateName(name: string): void {
  if (!name || new TextEncoder().encode(name).length > 200 || name.split(".").some((part) => !part)) {
    invalid("invalid gateway name");
  }
}
export interface Route {
  readonly name: string;
  readonly kind: "unary" | "serverStreaming";
  readonly controlVersion: number;
  readonly contractVersion: number;
  readonly input: string;
  readonly output: string;
}
export function validateRoute(route: Route): void {
  validateName(route.name);
  validateName(route.input);
  validateName(route.output);
  if (route.controlVersion !== CONTROL_VERSION || route.contractVersion !== CONTRACT_VERSION) {
    throw new GatewayFailure({ code: "INCOMPATIBLE", message: "unsupported protocol or contract version" });
  }
  if (route.kind !== "unary" && route.kind !== "serverStreaming") invalid("invalid method kind");
}
export function accepts(actual: Route, request: Route): Result<void> {
  if (
    actual.name !== request.name ||
    actual.kind !== request.kind ||
    actual.input !== request.input ||
    actual.output !== request.output ||
    actual.controlVersion !== request.controlVersion ||
    actual.contractVersion !== request.contractVersion
  )
    return failure("INCOMPATIBLE", "route contract mismatch");
  if (actual.kind !== "unary") return failure("WRONG_METHOD_KIND", "stream execution requires stream API");
  return success(undefined);
}
export type Backpressure = "ordered" | "coalesce" | "drop";
export interface EventDescriptor {
  readonly name: string;
  readonly policy: Backpressure;
}
export interface RegisteredRoute extends Route {
  readonly streamPolicy?: Partial<import("./stream.js").StreamPolicy>;
  readonly timeoutMs: number;
  readonly maxConcurrency: number;
}
export interface Manifest {
  readonly routes: readonly RegisteredRoute[];
  readonly events: readonly EventDescriptor[];
}
export interface Subscription {
  close(): void;
}
export interface Transport {
  stream?(
    route: Route,
    payload: Uint8Array,
    options?: import("./stream.js").StreamOptions,
  ): Promise<import("./stream.js").ResponseStream<Uint8Array>>;
  invoke(route: Route, payload: Uint8Array): Promise<Result<Uint8Array>>;
  subscribe(
    event: string,
    filter: Uint8Array | undefined,
    sink: EventSink,
    persistent?: boolean,
  ): Promise<Subscription>;
}
export type EventSink = (payload: Uint8Array) => void | Promise<void>;

File: gateway/rust/src/protocol.rs
use std::collections::HashSet;
use std::fmt;
use std::sync::Arc;

use serde::{Deserialize, Serialize};

pub const CONTROL_VERSION: u32 = 1;
pub const CONTRACT_VERSION: u32 = 1;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorCode {
    UnknownRoute,
    OwnerUnavailable,
    InvalidArgument,
    ConcurrencyFull,
    Timeout,
    Unauthorized,
    HandlerError,
    Conflict,
    Incompatible,
    WrongMethodKind,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct GatewayError {
    pub code: ErrorCode,
    pub message: String,
}

impl GatewayError {
    pub fn new(code: ErrorCode, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
        }
    }
}

impl fmt::Display for GatewayError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{:?}: {}", self.code, self.message)
    }
}
impl std::error::Error for GatewayError {}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum MethodKind {
    Unary,
    ServerStreaming,
}

/// Contract versions represent explicit breaking revisions, not descriptor hashes.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Route {
    pub name: String,
    pub kind: MethodKind,
    pub control_version: u32,
    pub contract_version: u32,
    pub input: String,
    pub output: String,
}

impl Route {
    pub fn validate(&self) -> Result<(), GatewayError> {
        validate_name(&self.name)?;
        if self.control_version != CONTROL_VERSION || self.contract_version != CONTRACT_VERSION {
            return Err(GatewayError::new(
                ErrorCode::Incompatible,
                "unsupported protocol or contract version",
            ));
        }
        validate_name(&self.input)?;
        validate_name(&self.output)
    }

    pub fn accepts(&self, requested: &Self) -> Result<(), GatewayError> {
        requested.validate()?;
        if self != requested {
            return Err(GatewayError::new(ErrorCode::Incompatible, "route contract mismatch"));
        }
        if self.kind != MethodKind::Unary {
            return Err(GatewayError::new(
                ErrorCode::WrongMethodKind,
                "stream execution requires stream API",
            ));
        }
        Ok(())
    }
}

pub fn validate_name(name: &str) -> Result<(), GatewayError> {
    if name.is_empty() || name.len() > 200 || name.split('.').any(str::is_empty) {
        return Err(GatewayError::new(ErrorCode::InvalidArgument, "invalid gateway name"));
    }
    Ok(())
}

/// Created by the host, never decoded from a business payload. Clients retain this
/// context across nested calls. Loading arbitrary code is not a sandbox boundary.
#[derive(Debug, Clone)]
pub struct CallContext(Arc<Permissions>);

#[derive(Debug)]
struct Permissions {
    caller: String,
    trusted: bool,
    invoke: HashSet<String>,
    subscribe: HashSet<String>,
}

impl CallContext {
    pub fn trusted(caller: impl Into<String>) -> Self {
        Self(Arc::new(Permissions {
            caller: caller.into(),
            trusted: true,
            invoke: HashSet::new(),
            subscribe: HashSet::new(),
        }))
    }

    pub fn restricted(caller: impl Into<String>, invoke: Vec<String>, subscribe: Vec<String>) -> Self {
        Self(Arc::new(Permissions {
            caller: caller.into(),
            trusted: false,
            invoke: invoke.into_iter().collect(),
            subscribe: subscribe.into_iter().collect(),
        }))
    }

    pub fn caller(&self) -> &str {
        &self.0.caller
    }

    pub fn authorize(&self, name: &str, event: bool) -> Result<(), GatewayError> {
        let allowed = if event { &self.0.subscribe } else { &self.0.invoke };
        if self.0.trusted || allowed.contains(name) {
            return Ok(());
        }
        Err(GatewayError::new(ErrorCode::Unauthorized, "capability not authorized"))
    }
}

pub type WireResult = Result<Vec<u8>, GatewayError>;

/// Owner announcement. Execution policy is descriptive at forwarding hosts.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RouteRegistration {
    #[serde(flatten)]
    pub route: Route,
    pub timeout_ms: u64,
    pub max_concurrency: usize,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub routes: Vec<RouteRegistration>,
    pub events: Vec<crate::event::EventDescriptor>,
}
